internal/cmd/oohelperd/http.go

```diff
@@ -102,13 +106,56 @@ func httpDo(ctx context.Context, config *httpConfig) {
 	reader := &io.LimitedReader{R: resp.Body, N: config.MaxAcceptableBody}
 	data, err := netxlite.ReadAllContext(ctx, reader)
 	ol.Stop(err)
+
+	h3Endpoint := ""
+	if !config.h3 {
+		h3Endpoint = getHTTP3Altsvc(resp)
+	}
+
 	config.Out <- ctrlHTTPResponse{
-		BodyLength: int64(len(data)),
-		Failure:    httpMapFailure(err),
-		StatusCode: int64(resp.StatusCode),
-		Headers:    headers,
-		Title:      measurexlite.WebGetTitle(string(data)),
+		BodyLength:   int64(len(data)),
+		DiscoveredH3: h3Endpoint,
+		Failure:      httpMapFailure(err),
+		StatusCode:   int64(resp.StatusCode),
+		Headers:      headers,
+		Title:        measurexlite.WebGetTitle(string(data)),
+	}
+}
+
+func getHTTP3Altsvc(resp *http.Response) string {
+	var responses []*http.Response
+	var request *http.Request
+
+	// go through chain of HTTP Request - Responses to find first Response
+	// TODO (kelmenhorst): this will not work for http:// URLs
+	for i := 0; i < 10; i++ {
+		responses = append(responses, resp)
+		request = resp.Request
+		if request == nil {
+			break
+		}
+		if request.Response == nil {
+			break
+		}
+		resp = request.Response
+	}
+	alt_svc := resp.Header.Get("Alt-Svc")
```

Comment:
> Otherwise, we should choose whether we want to loop over all the entries or whether it's enough for our goal to just process the first entry. The former is probably more robust and we should probably go for the former given that we're not writing one-off code such as websteps-illustrated.

Could you elaborate on why you think it suffices to only process the first comma-separated entry in the Alt-Svc header? Is there some convention to put `h3` first?

Reply:
> Could you elaborate on why you think it suffices to only process the first comma-separated entry in the Alt-Svc header? Is there some convention to put h3 first?

I don't think this convention is written anywhere and the resulting code would be less precise and probably fail in a bunch of cases. I think the ideal would be to walk through all the entries and match the first that has `h3` and then extract the information from it. Even if most websites put `h3` first as of now, future changes in how the internet works (maybe `h4` would be first in one year?) could mean that our code will break. And it will break a bit silently as in "suddenly we see a drop in h3 measurements and we don't understand why". So, maybe it's really worth the effort of writing the code in a way that, if any `h3` entry is there, we parse it. What do you think?

Reply:
>> For each entry, we should split at ;. However, it's optimistic to unconditionally take the [0] element: what if the list is empty? So, if there are less than 1 elements we should skip the entry.

I was wrong here. The usage of `strings.Strip` you're using is such that it cannot return a zero-length slice. Because I (and possibly others) would still be surprised by seeing `v[0]` after a function that returns `v []string`, I think we should make this fact explicit by adding `runtimex.Assert` with a comment that explains the assertion.

Reply:
I agree with both your suggestions, thank you!